waiter/src/waiter/auth/authentication.clj

```diff
@@ -201,3 +219,62 @@
         (handler request)))))
+
+(defn process-auth-expires-at-request
+  "Handler to allow a client to update it's knowledge of when user's cookie-based credentials expire.
+   Returns a json response containing the expires-at key containing the expiration time in UTC epoch seconds.
+   Relies on the metadata in the x-waiter-auth cookie."
+  [password {:keys [headers]}]
+  (let [decoded-auth-cookie (get-and-decode-auth-cookie-value headers password)
+        [auth-principal _ auth-metadata] decoded-auth-cookie
+        {:keys [expires-at]} auth-metadata
+        expires-at (or (when (decoded-auth-valid? decoded-auth-cookie) expires-at) 0)]
+    (log/info auth-principal "cookie expires at" expires-at)
+    (utils/attach-waiter-source (utils/clj->json-response {:expires-at expires-at
+                                                           :principal auth-principal}))))
+
+(defn process-auth-keep-alive-request
+  "Handler to eagerly trigger authentication workflow even if cookie has not yet expired.
+   This allows clients to pre-emptively refresh credentials before they expire.
+   Presence of done parameter is used to avoid infinite auth redirect loops and will return 204 No Content.
+   Invalid offset parameters result in a 400 error.
+   Missing offset or soon to expire cookie (based on offset) will trigger in the authentication workflow.
+   If cookie is expected to be live longer than offset value, return 204 No Content."
+  [password auth-handler {:keys [headers] :as request}]
+  (let [{:strs [done offset]} (-> request ru/query-params-request :query-params)
+        offset-parsed (utils/parse-int offset)
+        decoded-auth-cookie (get-and-decode-auth-cookie-value headers password)
+        [auth-principal _ cookie-metadata] decoded-auth-cookie
+        current-epoch-time (tc/to-epoch (t/now))
+        {:keys [expires-at]} cookie-metadata
+        ;; handle legacy cookies which will not have this value set
+        expires-at (or expires-at current-epoch-time)]
+    (log/info auth-principal "cookie expires at" expires-at "offset is" offset-parsed)
+    (cond
+      ;; avoid infinite redirect loop
+      done
+      (utils/attach-waiter-source {:status http-204-no-content})
+
+      ;; offset parameter provided, but cannot be parsed
+      (and offset (nil? offset-parsed))
+      (utils/clj->json-response {:message "Unable to parse offset parameter"
+                                 :parameter {:offset offset}}
+                                :status http-400-bad-request)
+
+      ;; offset parameter must be positive when provided
+      (and offset-parsed (not (pos? offset-parsed)))
+      (utils/clj->json-response {:message "Invalid offset parameter"
+                                 :parameter {:offset offset-parsed}}
+                                :status http-400-bad-request)
+
+      ;; trigger auth workflow if cookie has already expired or adding offset will cause it to expire
+      (or (nil? offset)
```

Comment:
The logic here doesn't match the comment. Why does eliding the offset/delta force a refresh?

Reply:
The issue wasn't with the docstring, it's with the comment on line 270. The `(nil? offset)` case doesn't seem to follow from the comment.

Reply:
Oops, updated the comment.